File: android/app/src/main/java/com/airbridge/app/data/relay/RelayMessageMapper.kt
package com.airbridge.app.data.relay

import com.airbridge.app.data.crypto.decodeBase64
import com.airbridge.app.data.crypto.encodeBase64
import com.airbridge.app.domain.BridgeChannel
import com.airbridge.app.domain.EncryptedEnvelope
import com.airbridge.app.domain.IncomingEncryptedEnvelope
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
sealed interface RelayClientMessage

@Serializable
data class RelayPingMessage(
    val type: String = "ping",
) : RelayClientMessage

@Serializable
data class RelaySendEnvelopeMessage(
    val type: String = "send_envelope",
    @SerialName("recipient_device_id")
    val recipientDeviceId: String,
    val channel: String,
    @SerialName("content_type")
    val contentType: String,
    val nonce: String,
    @SerialName("header_aad")
    val headerAad: String,
    val ciphertext: String,
) : RelayClientMessage

@Serializable
data class RelayAckEnvelopeMessage(
    val type: String = "ack_envelope",
    @SerialName("envelope_id")
    val envelopeId: String,
) : RelayClientMessage

@Serializable
data class RelayServerMessage(
    val type: String,
    val code: String? = null,
    val message: String? = null,
    @SerialName("device_id")
    val deviceId: String? = null,
    @SerialName("peer_device_id")
    val peerDeviceId: String? = null,
    @SerialName("envelope_id")
    val envelopeId: String? = null,
    @SerialName("sender_device_id")
    val senderDeviceId: String? = null,
    val channel: String? = null,
    @SerialName("content_type")
    val contentType: String? = null,
    val nonce: String? = null,
    @SerialName("header_aad")
    val headerAad: String? = null,
    val ciphertext: String? = null,
    @SerialName("created_at")
    val createdAt: String? = null,
    @SerialName("expires_at")
    val expiresAt: String? = null,
)

sealed interface RelayServerEvent {
    data class Connected(val deviceId: String, val peerDeviceId: String) : RelayServerEvent
    data object Pong : RelayServerEvent
    data class Envelope(val envelope: IncomingEncryptedEnvelope) : RelayServerEvent
    data class Error(val code: String, val message: String) : RelayServerEvent
    data class Closed(val code: Int, val reason: String) : RelayServerEvent
    data class Failure(val throwable: Throwable) : RelayServerEvent
}

fun EncryptedEnvelope.toWire(recipientDeviceId: String): RelaySendEnvelopeMessage {
    return RelaySendEnvelopeMessage(
        recipientDeviceId = recipientDeviceId,
        channel = channel.wireValue(),
        contentType = contentType,
        nonce = nonce.encodeBase64(),
        headerAad = headerAad.encodeBase64(),
        ciphertext = ciphertext.encodeBase64(),
    )
}

fun RelayServerMessage.toEvent(): RelayServerEvent = when (type) {
    "connected" -> RelayServerEvent.Connected(
        deviceId = requireNotNull(deviceId),
        peerDeviceId = requireNotNull(peerDeviceId),
    )
    "pong" -> RelayServerEvent.Pong
    "envelope" -> RelayServerEvent.Envelope(
        IncomingEncryptedEnvelope(
            envelopeId = requireNotNull(envelopeId),
            senderDeviceId = requireNotNull(senderDeviceId),
            channel = BridgeChannel.valueOf(requireNotNull(channel).uppercase()),
            contentType = requireNotNull(contentType),
            nonce = requireNotNull(nonce).decodeBase64(),
            headerAad = requireNotNull(headerAad).decodeBase64(),
            ciphertext = requireNotNull(ciphertext).decodeBase64(),
        ),
    )
    "error" -> RelayServerEvent.Error(
        code = code ?: "error",
        message = message ?: "relay websocket 오류가 발생했어요",
    )
    else -> RelayServerEvent.Error(
        code = "unknown_message_type",
        message = "알 수 없는 서버 메시지를 받았어요: $type",
    )
}

private fun BridgeChannel.wireValue(): String = when (this) {
    BridgeChannel.CLIPBOARD -> "clipboard"
    BridgeChannel.NOTIFICATION -> "notification"
}

File: mac/AirBridgeMac/Data/Relay/RelayMessageMapper.swift
import Foundation

enum RelayMessageMapperError: LocalizedError, Equatable {
    case invalidServerMessage
    case invalidEnvelope

    var errorDescription: String? {
        switch self {
        case .invalidServerMessage:
            return "The relay sent an invalid WebSocket message."
        case .invalidEnvelope:
            return "The relay sent an invalid envelope."
        }
    }
}

enum RelayClientMessage {
    case ping
    case sendEnvelope(
        recipientDeviceID: String,
        channel: RelayChannel,
        contentType: String,
        nonce: Data,
        headerAAD: Data,
        ciphertext: Data
    )
    case acknowledgeEnvelope(envelopeID: String)
}

enum RelayServerMessage: Equatable {
    case connected(deviceID: String, peerDeviceID: String)
    case pong
    case envelope(RelayEnvelope)
    case error(code: String, message: String)
}

private struct RelayClientMessageDTO: Encodable {
    let type: String
    let envelopeID: String?
    let recipientDeviceID: String?
    let channel: String?
    let contentType: String?
    let nonce: String?
    let headerAAD: String?
    let ciphertext: String?

    enum CodingKeys: String, CodingKey {
        case type
        case envelopeID = "envelope_id"
        case recipientDeviceID = "recipient_device_id"
        case channel
        case contentType = "content_type"
        case nonce
        case headerAAD = "header_aad"
        case ciphertext
    }
}

private struct RelayServerMessageDTO: Decodable {
    let type: String
    let code: String?
    let message: String?
    let deviceID: String?
    let peerDeviceID: String?
    let envelopeID: String?
    let senderDeviceID: String?
    let channel: String?
    let contentType: String?
    let nonce: String?
    let headerAAD: String?
    let ciphertext: String?
    let createdAt: Date?
    let expiresAt: Date?

    enum CodingKeys: String, CodingKey {
        case type
        case code
        case message
        case deviceID = "device_id"
        case peerDeviceID = "peer_device_id"
        case envelopeID = "envelope_id"
        case senderDeviceID = "sender_device_id"
        case channel
        case contentType = "content_type"
        case nonce
        case headerAAD = "header_aad"
        case ciphertext
        case createdAt = "created_at"
        case expiresAt = "expires_at"
    }
}

struct RelayMessageMapper {
    func encode(_ message: RelayClientMessage) throws -> String {
        let dto: RelayClientMessageDTO

        switch message {
        case .ping:
            dto = RelayClientMessageDTO(
                type: "ping",
                envelopeID: nil,
                recipientDeviceID: nil,
                channel: nil,
                contentType: nil,
                nonce: nil,
                headerAAD: nil,
                ciphertext: nil
            )
        case .acknowledgeEnvelope(let envelopeID):
            dto = RelayClientMessageDTO(
                type: "ack_envelope",
                envelopeID: envelopeID,
                recipientDeviceID: nil,
                channel: nil,
                contentType: nil,
                nonce: nil,
                headerAAD: nil,
                ciphertext: nil
            )
        case .sendEnvelope(
            let recipientDeviceID,
            let channel,
            let contentType,
            let nonce,
            let headerAAD,
            let ciphertext
        ):
            let validatedRecipientDeviceID = try RelayInputValidator.identifier(
                recipientDeviceID,
                field: "recipient_device_id"
            )
            let validatedContentType = try RelayInputValidator.envelope(
                contentType: contentType,
                nonce: nonce,
                headerAAD: headerAAD,
                ciphertext: ciphertext
            )

            dto = RelayClientMessageDTO(
                type: "send_envelope",
                envelopeID: nil,
                recipientDeviceID: validatedRecipientDeviceID,
                channel: channel.rawValue,
                contentType: validatedContentType,
                nonce: nonce.rawBase64EncodedString,
                headerAAD: headerAAD.rawBase64EncodedString,
                ciphertext: ciphertext.rawBase64EncodedString
            )
        }

        return String(decoding: try JSONEncoder.airBridge.encode(dto), as: UTF8.self)
    }

    func decode(_ text: String) throws -> RelayServerMessage {
        let dto = try JSONDecoder.airBridge.decode(RelayServerMessageDTO.self, from: Data(text.utf8))

        switch dto.type {
        case "connected":
            guard let deviceID = dto.deviceID,
                  let peerDeviceID = dto.peerDeviceID else {
                throw RelayMessageMapperError.invalidServerMessage
            }

            return .connected(
                deviceID: deviceID,
                peerDeviceID: peerDeviceID
            )
        case "pong":
            return .pong
        case "error":
            return .error(
                code: dto.code ?? "unknown_error",
                message: dto.message ?? "Unknown WebSocket error."
            )
        case "envelope":
            guard let envelopeID = dto.envelopeID,
                  let senderDeviceID = dto.senderDeviceID,
                  let channelRawValue = dto.channel,
                  let channel = RelayChannel(rawValue: channelRawValue),
                  let contentType = dto.contentType,
                  let nonceValue = dto.nonce,
                  let headerAADValue = dto.headerAAD,
                  let ciphertextValue = dto.ciphertext,
                  let createdAt = dto.createdAt,
                  let expiresAt = dto.expiresAt else {
                throw RelayMessageMapperError.invalidEnvelope
            }

            let nonce = try Data(rawBase64Encoded: nonceValue)
            let headerAAD = try Data(rawBase64Encoded: headerAADValue)
            let ciphertext = try Data(rawBase64Encoded: ciphertextValue)
            let validatedContentType = try RelayInputValidator.envelope(
                contentType: contentType,
                nonce: nonce,
                headerAAD: headerAAD,
                ciphertext: ciphertext
            )

            return .envelope(
                RelayEnvelope(
                    id: envelopeID,
                    senderDeviceID: senderDeviceID,
                    channel: channel,
                    contentType: validatedContentType,
                    nonce: nonce,
                    headerAAD: headerAAD,
                    ciphertext: ciphertext,
                    createdAt: createdAt,
                    expiresAt: expiresAt
                )
            )
        default:
            return .error(code: "unknown_message_type", message: "Unknown WebSocket message type.")
        }
    }
}
